src/__tests__/regressions.test.ts

```diff
@@ -0,0 +1,166 @@
+  // This regression-tests a user-state achievable throw where you open, then during
+  // connection metadata fetch, open, close, open, close and end up with an interlaced
+  // connection state as a result of handleClose resetting the connection status.
+  // CrosisError: Client entered wrong state during connect(); connected=false
+
+  let waitingPromise: Promise<FetchConnectionMetadataResult> | null = null;
+  const waitingMetadata = () => {
+    if (waitingPromise) {
+      return waitingPromise;
+    }
+
+    waitingPromise = new Promise<FetchConnectionMetadataResult>((resolve) => {
+      return getConnectionMetadata().then(resolve);
+    });
+    return waitingPromise;
+  };
+
+  let abortingPromise: Promise<FetchConnectionMetadataResult> | null = null;
+  const abortingMetadata = () => {
+    if (abortingPromise) {
+      return abortingPromise;
+    }
+
+    abortingPromise = new Promise<FetchConnectionMetadataResult>((resolve) => {
+      resolve({
+        error: FetchConnectionMetadataError.Aborted,
+      });
+    });
+    return abortingPromise;
+  };
+
+  const client = getClient<{ username: string }>(done);
+  const ctx = { username: 'zyzz' };
+
+  const params = {
+    fetchConnectionMetadata: getConnectionMetadata,
+    WebSocketClass: WebSocket,
+    context: ctx,
+  };
+  client.open(
+    { ...params, fetchConnectionMetadata: abortingMetadata },
+    wrapWithDone(done, () => {
+      throw new Error('Expected to never get here (abort+instant close).');
+    }),
+  );
+
+  client.close();
+
+  client.open(
+    // this is currently expected to return aborted, I think, but that should be
+    // encapsulated inside the client.
+    { ...params, fetchConnectionMetadata: waitingMetadata },
+    wrapWithDone(done, () => {
+      // we could possibly get here, it is maybe flaky to rely on the order of
+      // the promise resolutions (or at least, I can't trace the exact order
+      // I expect).
+      console.log('reached second client.open');
+    }),
+  );
+
+  await abortingPromise;
+
+  client.close();
+
+  await waitingPromise;
+
+  done();
+});
+
+concurrent("interlaced connections don't throw - abort after case", async (done) => {
+  // This isn't known to be user-achievable, but is roughly the same condition as the
+  // abort-during case, but with the abort after the final close... this case is probably
+  // more straightforward that it should've returned aborted, but it is still causing pain in production.
+  // CrosisError: Expected abort returned from fetchConnectionMetadata to be truthy when the controller aborts
+
+  let waitingPromise: Promise<FetchConnectionMetadataResult> | null = null;
+  const waitingMetadata = () => {
+    if (waitingPromise) {
+      return waitingPromise;
+    }
+
+    waitingPromise = new Promise<FetchConnectionMetadataResult>((resolve) => {
+      return getConnectionMetadata().then(resolve);
+    });
+    return waitingPromise;
+  };
```

Comment:
This one is less clear that it is an actual 'bug' (vs. just missed encapsulation) because if you replace it with the 'correct' implementation, everything works as expected. The error message is also clear that's the case.

```suggestion
  let waitingPromise: Promise<FetchConnectionMetadataResult> | null = null;
  const waitingMetadata = (abortSignal: AbortSignal) => {
    if (waitingPromise) {
      return waitingPromise;
    }

    waitingPromise = new Promise<FetchConnectionMetadataResult>((resolve) => {
      return getConnectionMetadata().then((r) => {
        if (abortSignal.aborted) {
          resolve({
            error: FetchConnectionMetadataError.Aborted,
          });
        }

        resolve(r);
      });
    });
    return waitingPromise;
  };
```

Reply:
Going to drop this one, per the point here: https://github.com/replit/crosis/pull/192